### ISSUE TITLE
__make_mdesc compiler warning

### ISSUE THREAD

**Issue by morrone** [ovis-hpc/ldms#1054]
ldms.c is triggering the following warning for me:

```
ldms.c: In function ‘__make_mdesc’:
ldms.c:1457:2: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
  strncpy(vd->vd_name_unit, md->name, name_len);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ldms.c:1456:13: note: length computed here
  name_len = strlen(md->name) + 1;
             ^~~~~~~~~~~~~~~~
ldms.c:1460:3: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
   strncpy(&vd->vd_name_unit[name_len], md->unit, unit_len);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ldms.c:1459:14: note: length computed here
   unit_len = strlen(md->unit) + 1;
              ^~~~~~~~~~~~~~~~
```

This warning seems valid to me. There doesn't seem to be checking to make sure that the destination is large enough. But I don't understand the macro magic going on here. I am not clear on how one would know what the available size of the destination is to add that check.

**Comment by tom95858** [comment]
We've been going around and around on this with @baallan. We really don't want to be using strncpy in my opinion because it is much less efficient than strcpy; we added all this because a static analysis tool was complaining about the use of strcpy. Most people don't realize this but strncpy not only validates the length, but zero fills the resulting size if the source len is less than the target len.

**Comment by morrone** [comment]
Sure. I was going to change the length option of strncpy() to essentially:

```
    len = min(dest_len, src_len)
```

That does avoid the extra zeros.

But then I realized that vd_name_unit_len is set at the _end_ of __make_mdesc, so I wasn't sure that I could trust it to have the a useful value when the function is entered. And walking up the call stack to see where that buffer comes from and the size allocated for vd_name_unit, I hit magic-macro-land and abandoned ship. :)

**Comment by baallan** [comment]
One good reason for the strncpy behavior is the case where the entire buffer (including the 'excessively zeroed' tail) is going to go over the wire and otherwise expose data that potentially should not be exposed.

When we are better at keeping track of input string lengths and calling strlen on the unknown string piece exactly once (which tends to expand argument lists with lengths), we can/should of course safely call strcpy and any decent audit tool will be able to prove to itself (across function call boundaries) that we are ok and have earned the improved performance. Fortify is once such audit tool; i cannot vouch for the accuracy of recent clang or recent gcc string audits.

**Comment by tom95858** [comment]
@baallan, @morrone strncpy only applies to byte_array metric types, not other primitive array types such as double, uint64, etc... Initially all set memory is bzero'd, however, there is the case where set memory is re-used when it is deleted and it's underlying set memory is reused by another set with potentially different uid/gid. Currently only the set's meta-data portion of the metric set is bzero'd; therefore, the data portion of a metric set could potentially contain the contents of a previous metric set. I think the correct fix is to bzero both the meta-data and the data portion of a metric set in ldms_set_create. In that case, the data portion is initialized and there is no danger of accidentally exposing metric data from another metric set owned by a different user. In any event, we are not talking about the classic ploy of getting stack data on the wire.

**Comment by tom95858** [comment]
@baallan, @morrone please reopen as required.

**Comment by baallan** [comment]
I'll accept @tom95858 proposed bzero solution in set_create, but afaict on top of tree it has not been implemented. It's possible i'm missing some equivalent in the code....

**Comment by morrone** [comment]
I was just looking at this again. So here are my latest thoughts:

I think the discussion about strcpy() vs strncpy() is kind of beside the point.

What this compiler warning is telling us is that the __make_mdesc() code is appearing to blindly write into vd->vd_name_unit without regards to whether there is sufficient memory at that location for the write. And from a programmer's prospective, one has to agree that there seems to be no checking here. There seems to be no _way_ for __mdesc_mdesc() to know what it is doing is safe. This looks like a bit of a sketchy programming practice, and _that_ is really what gcc is (correctly, I'd say) pointing out.

In fact, ldms_set_create() in general seems to just blindly build the set starting with:

```
  vd = get_first_metric_desc(meta);
```

And then just assumes infinite memory returned by get_first_metric_desc(). There doesn't seem to be any tracking of how much space is available in the pointer returned from get_first_metrict_desc(). It seems like we almost certainly _should_ be tracking the remaining memory.

And if we know the remaining memory, we could pass that number to __make_mdesc(). And then __make_mdesc() could do the proper checking to make sure it never over flows at write operation (again, regardless of whether it uses strcpy() or strncpy()).

In summary, getting rid of strncpy() might get rid of the compiler error, but it doesn't fix the bad practice of writing into a memory buffer without regards (seemingly) to how much memory is available in said buffer.

**Comment by emdonat** [comment]
destination size rather than source size will be used in 4.5.1
